server/sonar-server/src/main/java/org/sonar/server/computation/step/QualityGateLoadingStep.java

```diff
@@ -77,16 +75,22 @@ private void executeForProject(Component project) {
 
     if (qualityGateSetting == null || StringUtils.isBlank(qualityGateSetting)) {
       LOGGER.debug("No quality gate is configured for project " + projectKey);
-      qualityGateHolder.setQualityGate(DEFAULT_QUALITY_GATE);
+      qualityGateHolder.setNoQualityGate();
       return;
     }
 
     try {
       long qualityGateId = Long.parseLong(qualityGateSetting.trim());
-      qualityGateHolder.setQualityGate(qualityGateService.findById(qualityGateId).or(DEFAULT_QUALITY_GATE));
+      Optional<QualityGate> qualityGate = qualityGateService.findById(qualityGateId);
+      if (qualityGate.isPresent()) {
+        qualityGateHolder.setQualityGate(qualityGate.get());
+      }
+      else {
+        qualityGateHolder.setNoQualityGate();
+      }
     } catch (NumberFormatException e) {
       LOGGER.error(String.format("Unsupported value in property %s, using empty Quality Gate", PROPERTY_QUALITY_GATE));
-      qualityGateHolder.setQualityGate(DEFAULT_QUALITY_GATE);
+      qualityGateHolder.setNoQualityGate();
```

Comment:
you're right, I'll throw an exception, this can not happen unless user modified the DB directly or something like that